### DISCUSSION TITLE
Scroll to keep currently focused control visible

### DISCUSSION THREAD

**Discussion by dasien** [tui-cs/Terminal.Gui#3452] [Q&A] [answered]
I am curious how i should go about keeping the currently focused control visible inside a `ScrollView`.  If the user is pressing the `Tab` key to move between controls, and they go beyond the visible part of the screen, i want the `ScrollView` to scroll down so that the next control (the one that just got focus) is visible.

Is there a way to trigger this behavior automatically or is this something that needs to be handled when the Tab key is pressed?

Thanks!
Brian

**Comment by dasien** [comment]
Still trying to figure this one out.  What does `KeepContentAlwaysInViewport` do?  I was hoping this would make sure the focused control was visible in the viewport, but it seems to have no effect.

**Reply by tig** [reply]
> Still trying to figure this one out. What does `KeepContentAlwaysInViewport` do? I was hoping this would make sure the focused control was visible in the viewport, but it seems to have no effect.

The name of `KeepContentAlwaysInViewport` is misleading. What it really does is ensure that when scrolling vertically, the last row of the content will always be visible and when scrolling horizontally the last column will always be visible.

To do what you're trying to do you'll need to

- For each Subview, subscribe to the `Enter` event. 
- In the handler, determine the position and location of the view relative to the viewport.
- Set the position of the ScrollView such that the view is fully visible.

In simple cases, this will be simple. However, you'll soon find that there are ton of edge cases and nuances. For example, should the view in question be positioned after adusting the position such that it is JUST visible (it's bottom & right sides are flush with the bottom-right of the viewport)? Or should it be inset by some factor? What if the view is larger than the viewport? Etc...

I want to enable this automatically in v2 given `View` now has built-in scrolling capability and we're deprecating `ScrollView`. But it's a Hard problem. Solvable, but not easily.

**Reply by tig** [reply, marked as answer]
FWIW, In v2, the equivalent of `KeepContentAlwaysInViewport ` (again, given ScrollView is soon to be deprecated and removed) is `AllowLocationGreaterThanContentSize`, `AllowXGreaterThanContentWidth`, and `AllowYGreaterThanContentHeight`.

See https://gui-cs.github.io/Terminal.GuiV2Docs/api/Terminal.Gui.ViewportSettings.html

**Comment by dasien** [comment]
Thank you for the reply!  i was on the same track with the bullets you have above, so i will tinker with tracking the bounds of the currently viewable section of the `ScrollView`.

**Reply by dasien** [reply]
Sorry @tig one more semi-related question.  It seems like it should be easy, but how do you get the Y value of a control as an integer?  I see that it is a `Pos`, and that you can set the X or Y as an int, but i don't see how to get the int value back out of a `View.Y` once it is set.

**Reply by tig** [reply]
`view.Frame.X`
